### PR TITLE
Nginx

### PR DESCRIPTION
**Pull request by LenaigKaliou**
NGINX V1 for migration

### CODE REVIEW

File: NGINX/nginx/ingest/parser.yml [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    external:
+      name: json.parse-json
+    filter: '{{original.message | re_match("\{(?:[^{}]|())*\}")}}'
+  - name: set_json_fields
+    filter: '{{original.message | re_match("\{(?:[^{}]|())*\}")}}'
+  - name: set_grok_fields
+    filter: '{{original.message | re_match("\{(?:[^{}]|())*\}") == False}}'
+
+stages:
+  set_json_fields:
```
Comment (squioc) [inline NGINX/nginx/ingest/parser.yml:23]: Suggesting added this constant ECS fields (maybe `event.type` should be adapted according to the original event: is an access log or an error?)
```suggestion
   set_ecs_fields:
    actions:
      - set:
          event.kind: 'event'
          event.category: ['web']
          event.type: ['access']
          event.dataset: 'access'
          observer.product: 'nginx'
          observer.type: 'WEB server'
          observer.vendor: 'F5'
    set_json_fields:
```

File: NGINX/nginx/tests/json_delete.json [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  },
+  "expected": {
+    "message": "{\"server_name\":\"rdp.acme.com\",\"server_addr\":\"5.6.7.8\",\"remote_addr\":\"1.2.3.4\",\"remote_user\":\"\",\"time_local\":\"02/Jun/2020:13:07:18 +0200\",\"request\":\"DELETE /blobstore/aaa/bbb/530fd9ee-23af-4d4e-a9e4-0b7280ace286/document.tgz HTTP/1.1\",\"status\":\"204\",\"body_bytes_sent\":\"0\",\"http_referer\":\"\",\"http_user_agent\":\"Mozilla/5.0 (Windows NT 6.1; Trident/7.0; rv:11.0) like Gecko\",\"proxy_add_x_forwarded_for\":\"1.2.3.4\",\"request_time\":\"0.007\",\"ssl_protocol\":\"\",\"ssl_cipher\":\"\",\"sent_http_content_type\":\"\"}",
+    "action": {
+      "name": "DELETE",
+      "outcome": "success",
+      "properties": [
+        {
+          "time_local": "02/Jun/2020:13:07:18 +0200",
+          "proxy_add_x_forwarded_for": "1.2.3.4",
```
Comment (squioc) [inline NGINX/nginx/tests/json_delete.json:19]: I suggest using `network.forward_ip` for this field

File: NGINX/nginx/tests/json_delete.json [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+          "request_time": "0.007"
```
Comment (squioc) [inline NGINX/nginx/tests/json_delete.json:20]: I suggest using `event.duration` (be carrefully, `event.duration` is in nanoseconds)

File: NGINX/nginx/tests/extended_emptybyte.json [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      "response": {
+        "status_code": 200
+      },
+      "version": "1.1"
+    },
+    "source": {
+      "address": "1.2.3.4",
+      "ip": "1.2.3.4",
+      "user": {
+        "name": "-"
```
Comment (squioc) [inline NGINX/nginx/tests/extended_emptybyte.json:34]: I suggest filtering `-` values

File: NGINX/nginx/tests/extended_emptybyte.json [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      "kind": "access"
+    },
+    "destination": {
+      "address": "rdp.acme.com",
+      "domain": "rdp.acme.com"
+    },
+    "http": {
+      "request": {
+        "method": "POST",
+        "referrer": "-"
```
Comment (squioc) [inline NGINX/nginx/tests/extended_emptybyte.json:23]: I suggest filtering `-` values

File: NGINX/nginx/tests/json_get_failed_graylog.json [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  },
+  "expected": {
+    "message": "{ \"timestamp\": \"2020-06-04T11:17:10+02:00\", \"remote_addr\": \"1.2.3.4\", \"body_bytes_sent\": 36, \"request_time\": 0.003, \"response_status\": 404, \"request\": \"GET /rest/plugins/1.0/com.atlassian.jira.plugins.jira-slack-server-integration-plugin-key/media/plugin-icon HTTP/1.1\", \"request_method\": \"GET\", \"host\": \"rdp.acme.com\", \"http_scheme\": \"https\", \"upstream_cache_status\": \"-\", \"upstream_addr\": \"9.8.7.6:38080\", \"http_x_forwarded_for\": \"-\", \"http_referrer\": \"https://rdp.acme.com/plugins/servlet/project-config/PVC/summary\", \"http_user_agent\": \"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Safari/537.36\", \"http_version\": \"HTTP/1.1\", \"nginx_access\": true }",
+    "action": {
+      "name": "GET",
+      "outcome": "failure",
+      "properties": [
+        {
+          "timestamp": "2020-06-04T11:17:10+02:00",
```
Comment (squioc) [inline NGINX/nginx/tests/json_get_failed_graylog.json:18]: I suggest using this field to compute `@timestamp`

File: NGINX/nginx/tests/json_get_failed_graylog.json [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+          "request_time": "0.003",
+          "http_scheme": "https",
```
Comment (squioc) [inline NGINX/nginx/tests/json_get_failed_graylog.json:20]: I suggest using `network.protocol`

File: NGINX/nginx/tests/json_get_failed_graylog.json [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+          "nginx_access": "True"
```
Comment (squioc) [inline NGINX/nginx/tests/json_get_failed_graylog.json:21]: I suggest not using this field, except, maybe to compute `event.type`

**Comment by LenaigKaliou** [comment]
- Merged json and grok fields in one global field
- Made the fixes highlighted in the first version
- Standardized timestamp in rfc3339

**Comment by LenaigKaliou** [comment]
- Added timestamp for access log
- Added some info in url fields for error logs

File: NGINX/nginx/ingest/parser.yml [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+                failure
+              {%- endif -%}
+            {%- endif -%}
+        filter: '{{original.message | re_match("\{(?:[^{}]|())*\}")}}'
+      - translate:
+        dictionary:
+          'GET': ['access']
+          'DELETE': ['deletion']
+        mapping:
+          action.name: event.type
```
Comment (squioc) [inline NGINX/nginx/ingest/parser.yml:163]: Event with category `web` only accept type `access`, `error` or `info` (cf [doc](https://www.elastic.co/guide/en/ecs/8.6/ecs-allowed-values-event-category.html#ecs-event-category-web)).
I suggest deleting this step as we have the step line 170 that will set the `event.type`

File: NGINX/nginx/ingest/parser.yml [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+            {%- set min = min-z -%}
+            {%- if min < 10 -%}
+              {%- set min = min | string() -%}
+              {%- set min = '0' + min -%}
+            {%- endif -%}
+            {%- if monthnum < 10 -%}
+              {{year}}-0{{monthnum}}-{{elt[0]}}T{{hour}}:{{min}}:{{sec}}.000000Z
+            {%- else -%}
+              {{year}}-{{monthnum}}-{{elt[0]}}T{{hour}}:{{min}}:{{sec}}.000000Z
+            {%- endif -%}
```
Comment (squioc) [inline NGINX/nginx/ingest/parser.yml:60]: Why not use the stage `date.parse` with format '%d/%b/%Y:%H:%M:%S %z' to parse the time_local?

**Review by squioc — CHANGES_REQUESTED** [review]

**Comment by LenaigKaliou** [comment]
- Made fixes on dates
- Fixes event.type for json logs (line 170 wasn't operational for those)

**Review by squioc — APPROVED** [review]

**Pull request merged**